### PR TITLE
module 3

### PR DESCRIPTION
**Pull request by Darmio**

### CODE REVIEW

File: service/src/main/java/com/epam/esm/service/dto/update/TagUpdate.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    public TagUpdate(String name) {
+        this.name = name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getName() {
```
Comment (Shchava) [inline service/src/main/java/com/epam/esm/service/dto/update/TagUpdate.java:29]: why not use lombok here ?

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    public GiftCertificate getById(long id) {
+        return giftCertificateDao.findById(id)
+                .orElseThrow(() -> new NotFoundException("Not found gift certificate by id(" + id + ")", NOT_FOUND_GIFT_CERTIFICATES));
+    }
+
+    @Transactional
+    @Override
+    public GiftCertificate create(GiftCertificateCreate dto) {
+        try {
+            List<TagDto> createTags = dto.getTags().stream()
```
Comment (Shchava) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:64]: it is used only for other stream, you can  just continue writing logic from newCreateTags

File: service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            filters.add(Filter.builder()
+                    .field(FIELD.DESCRIPTION.toString())
+                    .operator(Filter.QueryOperator.LIKE)
+                    .value(searchParams.getDescription())
+                    .build());
+        }
+
+        if (searchParams.getPart() != null && !searchParams.getPart().isBlank()) {
+            filters.add(Filter.builder()
+                    // .field("part")
```
Comment (Shchava) [inline service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java:48]: please, remove it

**Review by Shchava — APPROVED** [review]

File: model/src/main/resources/db_queries.sql [hunk 4]
```diff
@@ -0,0 +1,40 @@
+-- DROP DATABASE rest_api;
+
+CREATE DATABASE IF NOT EXISTS rest_api;
```
[the remark this answers is not in the document]
Comment (Darmio) [reply model/src/main/resources/db_queries.sql:3]: this is from the previous module, queries  by hibernate used

File: service/src/main/java/com/epam/esm/service/dto/update/TagUpdate.java (hunk 1 above)
Comment (Darmio) [reply service/src/main/java/com/epam/esm/service/dto/update/TagUpdate.java:29]: Used before but not catched endpoint. I think its some problem with lombok

File: service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            default:
+                throw new RuntimeException("Operation not supported yet");
+        }
+    }
+
+    private static Specification<GiftCertificate> getSpecificationFromFilters(List<Filter> filters) {
+        if (filters.size() == 0) {
+            return where(null);
+        }
+        Specification<GiftCertificate> specification = where(createSpecification(filters.remove(0)));
```
[the remark this answers is not in the document]
Comment (Darmio) [reply service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java:110]: fixed

File: service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java (hunk 3 above)
Comment (Darmio) [reply service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java:48]: fixed

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 2 above)
Comment (Darmio) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:64]: fixed

File: service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            return Double.valueOf(value);
+        } else if (fieldType.isAssignableFrom(Long.class)) {
+            return Long.valueOf(value);
+        } else if (Enum.class.isAssignableFrom(fieldType)) {
+            return Enum.valueOf(fieldType, value);
+        }
+        return null;
+    }
+
+    private enum FIELD {
```
[the remark this answers is not in the document]
Comment (Darmio) [reply service/src/main/java/com/epam/esm/service/specification/GiftCertificateSpecification.java:128]: fixed

File: service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@Rollback()
```
[the remark this answers is not in the document]
Comment (Darmio) [reply service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:38]: fixed

File: service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        assertEquals(1, result.getContent().size());
+        assertEquals(expected, result);
+        verify(giftCertificateDao, times(0)).findAll();
+        verify(giftCertificateDao).findAll(any(Specification.class), any(Pageable.class));
+
+        verifyNoMoreInteractions(giftCertificateDao);
+    }
+
+    @Test
+    void getAllByParamsNameFalse() {
```
[the remark this answers is not in the document]
Comment (Darmio) [reply service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:95]: fixed

**Pull request merged**